File: Player/src/main/java/ru/testsimpleapps/coloraudioplayer/ui/adapters/ExplorerFilesAdapter.java
package ru.testsimpleapps.coloraudioplayer.ui.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import ru.testsimpleapps.coloraudioplayer.R;
import ru.testsimpleapps.coloraudioplayer.managers.explorer.Data.ItemData;
import ru.testsimpleapps.coloraudioplayer.managers.explorer.MediaExplorerManager;
import ru.testsimpleapps.coloraudioplayer.managers.tools.TimeTool;


public class ExplorerFilesAdapter extends BaseAdapter<ItemData> {

    private static final int TYPE_HEADER = 0;
    private static final int TYPE_ITEM = 1;
    private final Context mContext;
    private long mTotalTime = -1L;

    public ExplorerFilesAdapter(@NonNull Context context) {
        super();
        mContext = context;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        switch (i) {
            case TYPE_HEADER: {
                final View viewItem = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.explorer_header_files, viewGroup, false);
                return new ViewHolderHeader(viewItem);
            }

            case TYPE_ITEM: {
                final View viewItem = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.explorer_file, viewGroup, false);
                return new ViewHolderItem(viewItem);
            }

            default:
                return null;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int i) {
        if (viewHolder instanceof ViewHolderItem) {
            final ViewHolderItem mViewHolder = (ViewHolderItem) viewHolder;
            final ItemData itemData = getItem(i);
            mViewHolder.mCheckFile.setTag(ViewHolderItem.TAG_CHECK, i);
            mViewHolder.mDurationFile.setText(TimeTool.getDuration(itemData.getDuration()));
            mViewHolder.mNameFile.setText(String.valueOf(itemData.getName()));
            mViewHolder.mCheckFile.setChecked(itemData.isChecked());
        } else if (viewHolder instanceof ViewHolderHeader) {
            final ViewHolderHeader mViewHolder = (ViewHolderHeader) viewHolder;
            mViewHolder.mCountFiles.setText(String.valueOf(getItemCount() - 1));
            mViewHolder.mTotalTime.setText(TimeTool.getDuration(mTotalTime));
        }
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + 1;
    }

    @Override
    public ItemData getItem(int index) {
        return super.getItem(index - 1);
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return TYPE_HEADER;
        }
        return TYPE_ITEM;
    }

    public void setItems(final List<ItemData> list) {
        super.setItems(list);
        mTotalTime = countTotalTime();
    }

    /*
    * Count total time for current folder/album/artist only one time
    * */
    private long countTotalTime() {
        long totalTime = 0;
        if (mList != null) {
            for (ItemData item : mList) {
                totalTime += item.getDuration();
            }
        }

        return totalTime;
    }

    protected class ViewHolderItem extends RecyclerView.ViewHolder {

        public static final int TAG_CHECK = 10 << 24;

        @BindView(R.id.explorer_file_duration) TextView mDurationFile;
        @BindView(R.id.explorer_file_name) TextView mNameFile;
        @BindView(R.id.explorer_file_check) CheckBox mCheckFile;

        public ViewHolderItem(final View view) {
            super(view);
            ButterKnife.bind(this, view);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mOnItemClickListener != null) {
                        mOnItemClickListener.onItemClick(view, getLayoutPosition());
                    }
                }
            });
        }

        @OnCheckedChanged(R.id.explorer_file_check)
        protected void onChecked(final CheckBox checkBox) {
            final int position = (int) checkBox.getTag(TAG_CHECK);
            final ItemData itemData = getItem(position);
            itemData.setChecked(checkBox.isChecked());
        }
    }

    protected class ViewHolderHeader extends RecyclerView.ViewHolder {
        @BindView(R.id.explorer_header_files_count_files_value) TextView mCountFiles;
        @BindView(R.id.explorer_header_files_total_time_value) TextView mTotalTime;

        public ViewHolderHeader(final View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

}

File: player/src/main/java/ru/testsimpleapps/coloraudioplayer/ui/adapters/ExplorerFilesAdapter.kt
package ru.testsimpleapps.coloraudioplayer.ui.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import kotlinx.android.synthetic.main.explorer_file.view.*
import kotlinx.android.synthetic.main.explorer_header_files.view.*
import ru.testsimpleapps.coloraudioplayer.R
import ru.testsimpleapps.coloraudioplayer.managers.explorer.Data.ItemData
import ru.testsimpleapps.coloraudioplayer.managers.tools.TimeTool


class ExplorerFilesAdapter(private val mContext: Context) : BaseListAdapter<ItemData>() {

    companion object {
        val TAG_CHECK = 10 shl 24
    }

    private var mTotalTime = -1L

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): RecyclerView.ViewHolder? {
        when (i) {
            BaseAdapter.TYPE_HEADER -> {
                val viewItem = LayoutInflater.from(viewGroup.context).inflate(R.layout.explorer_header_files, viewGroup, false)
                return ViewHolderHeader(viewItem)
            }

            BaseAdapter.TYPE_ITEM -> {
                val viewItem = LayoutInflater.from(viewGroup.context).inflate(R.layout.explorer_file, viewGroup, false)
                return ViewHolderItem(viewItem)
            }

            else -> return null
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, i: Int) {
        if (viewHolder is ViewHolderItem) {
            val itemData = getItem(i)
            viewHolder.itemView.explorer_file_image!!.setTag(TAG_CHECK, i)
            viewHolder.itemView.explorer_file_duration!!.text = TimeTool.getDuration(itemData!!.duration)
            viewHolder.itemView.explorer_file_name!!.text = itemData.name.toString()
            viewHolder.itemView.explorer_file_check!!.isChecked = itemData.isChecked
        } else if (viewHolder is ViewHolderHeader) {
            viewHolder.itemView.explorer_header_files_count_folders_label!!.text = (itemCount - 1).toString()
            viewHolder.itemView.explorer_header_files_total_time_label!!.text = TimeTool.getDuration(mTotalTime)
        }
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + 1
    }

    override fun getItem(index: Int): ItemData? {
        return super.getItem(index - 1)
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) {
            BaseAdapter.TYPE_HEADER
        } else BaseAdapter.TYPE_ITEM
    }

    override fun setItems(list: List<ItemData>?) {
        super.setItems(list)
        mTotalTime = countTotalTime()
    }

    /*
    * Count total time for current folder/album/artist only one time
    * */
    private fun countTotalTime(): Long {
        var totalTime: Long = 0
        if (mList != null) {
            for (item in mList!!) {
                totalTime += item.duration
            }
        }

        return totalTime
    }

    inner class ViewHolderItem(view: View) : RecyclerView.ViewHolder(view) {

        init {
            view.setOnClickListener {
                if (mOnItemClickListener != null) {
                    mOnItemClickListener.onItemClick(view, layoutPosition)
                }
            }
        }
    }

    protected inner class ViewHolderHeader(view: View) : RecyclerView.ViewHolder(view)

}
